feat(client): ✨ Create view Account page and functionality

zaverbac opened the pull request.

zaverbac committed:
refactor(client): 🔥 Remove vite.config.timestamp.jms

Remove vite.config.timestamp.jms since file was uneeded

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
feat(client): ✨ Add link google account feature

Add link google account feature and error checking to see if user already had google account linked

zaverbac committed:
feat(client): ✨ Add link microsoft account functionality and error checking

Add complete functionality for a user to link a microsoft provider to their existing account along with checking if one is already linked and displaying an appropriate error

zaverbac committed:
feat(client): ✨ Add set password functionality

Add set password functionality to account page

zaverbac committed:
feat(client): ✨ Add appropriate error messages for linked accounts; Add user email and display name on page

Add error messages telling user that they have already linked an account with that provider to their current account and add users email and display name on account settings page

zaverbac committed:
fix(client): 🐛 Fix error checking bug

Fix error checking bug where first error that shows up is the only error that ever gets shown

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +54,165 @@
+
+  function linkGoogleAccount() {
+    try {
+      linkWithGoogle();
+      googleLinked = true;
```
This displays I successfully linked google account even though i only clicked the button.

zaverbac replied:
Fixed

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +54,165 @@
+
+  function linkMicrosoftAccount() {
+    try {
+      linkWithMicrosoft();
+      microsoftLinked = true;
```
This displays I successfully linked microsoft account even though i only clicked the button.

gchamb commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +42,81 @@ export async function loginAnonymous() {
+      }
+    });
+
+    linkWithPopup(user, google);
+    return;
```
probably want to return `linkWithPopup` then awaited when you use it

gchamb replied:
probably want to return `linkWithPopup` then awaited when you use it

nstringham replied:
fixed

gchamb commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +42,81 @@ export async function loginAnonymous() {
+    });
+
+    linkWithPopup(user, microsoft);
+    return;
```
probably want to return linkWithPopup then awaited when you use it

nstringham replied:
fixed

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +54,165 @@
+        <Label>Link Account Options</Label>
+      </Button>
+    </div>
+    <Dialog bind:open={showOptions} aria-labelledby="link-options-title" aria-describedby="link-options-content">
```
Dialog doesn't close after you link the google/microsoft acc

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +54,165 @@
-  <main>
+  <main class="settings-wrapper">
+    <h2>Account Settings</h2>
+    <p id="email">Email: {getEmail()}</p>
```
probably want to use authStore to get the users email because it isn't updating after i link my google acc

zaverbac replied:
Using authStore to get email

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +54,165 @@
+            <p class="success">Successfully Linked Microsoft Account</p>
+          {/if}
+        </div>
+        <h3>Set Password</h3>
```
works but doesn't closes the dialog after setting the password

zaverbac replied:
Closes dialog

gchamb commented:
On line 114 in auth.ts where is getDisplayName function being used for?

gchamb reviewed (dismissed):
see changes above

zaverbac committed:
refactor(client): ♻️ Change Link functions

Change link account functions to return popup and with linkPassword returns update password and awaits them when called in settings page

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
fix(client): 🐛 Fix weak password error bug

Fix bug where weak password error not showing up when thrown

zaverbac committed:
fix(client): 🐛 Fix email not updating

Fix email not updating and either displaying as null or an previous email before user linked new account

zaverbac committed:
refactor(client): 🔥 Remove getEmail function

Remove getEmail function as it was a redundant method

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +53,166 @@
+      await linkWithGoogle();
+      googleLinked = true;
+      showOptions = false;
+      return;
```
you dont need to return

nstringham replied:
still an issue

nstringham replied:
I think this is done now

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +53,166 @@
+    } catch (err) {
+      errorMsg = err instanceof Error ? err.message : String(err);
+      outputErrMsg();
+      return;
```
no need to return

nstringham replied:
still an issue

nstringham replied:
I think this is done now

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +53,166 @@
+      await linkWithMicrosoft();
+      microsoftLinked = true;
+      showOptions = false;
+      return;
```
no need to return

nstringham replied:
still an issue

nstringham replied:
I think this is done now

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +53,166 @@
+    } catch (err) {
+      errorMsg = err instanceof Error ? err.message : String(err);
+      outputErrMsg();
+      return;
```
no need to return

nstringham replied:
still an issue

nstringham replied:
I think this is done now

gchamb commented on client/src/routes/settings/+page.svelte:
```diff
@@ -21,11 +53,166 @@
+    } catch (err) {
+      errorMsg = err instanceof Error ? err.message : String(err);
+      outputErrMsg();
+      return;
```
no need to return

nstringham replied:
I think this is done now

zaverbac committed:
refactor(client): 🔥 Remove return fron link account functions

Remove return from link account functions as it is redundant

gchamb commented:
refresh page so updated link google/microsoft acc can update the email field

zaverbac committed:
fix(client): 🐛 Fix user email not updating on page

Fix user email not updating on account page by reloading page after accounts are successfully linked

zaverbac committed:
refactor(client): 🔥 Remove window.location.reload()

Remove window.location.reload() from linkPassword as there is no point in refreshing the page in that particular instance

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -8,8 +8,13 @@ import {
+  updatePassword,
 } from "firebase/auth";
 
+const google = new GoogleAuthProvider();
+const microsoft = new OAuthProvider("microsoft.com");
```
lets do these inside of a function

nstringham replied:
lets do these inside of a function

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +42,56 @@ export async function loginAnonymous() {
+    });
+    return linkWithPopup(user, google);
+  } else {
+    return user;
```
I feel like this should be an error

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +42,56 @@ export async function loginAnonymous() {
+
+    return linkWithPopup(user, microsoft);
+  } else {
+    return user;
```
I feel like this should be an error

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +42,56 @@ export async function loginAnonymous() {
+  if (user != null) {
+    return updatePassword(user, password);
+  } else {
+    return user;
```
I feel like this should be an error

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -62,9 +243,51 @@
+<style>
+  .settings-wrapper {
+    display: grid;
+    grid-template-rows: repeat(5, 50px [row-start]);
```
what?

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -1,31 +1,212 @@
+            type={showPassword ? "text" : "password"}
+            bind:value={confirmPassword}
+            required
+          />
```
dont need this because we have a show password button

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -1,31 +1,212 @@
+        passErr = "Password should be at least 6 characters";
+        return;
+      default:
+        errorMsg = "An unexpected error has occured";
```
lets show the error message instead of just saying unknown

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -1,31 +1,212 @@
+      <Title id="link-options-title">Link Account Options</Title>
+      <Content id="link-options-content">
+        <div class="signin-buttons">
+          <Button id="sign-in-with-google" variant="raised" on:click={linkGoogleAccount}>
+            <Icon component={Svg} viewBox="0 0 48 48">
+              {#each googleSvgPaths as path}
+                <path fill={path.color} d={path.path} />
+              {/each}
+            </Icon>
+            <Label>Sign in with Google</Label>
+          </Button>
```
this button should only be here if you don't alread have google account linked

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -1,31 +1,212 @@
+          {:else if googleLinked && googleErr === ""}
+            <p class="success">Successfully Linked Google Account</p>
+          {/if}
+          <Button id="sign-in-with-microsoft" variant="raised" on:click={linkMicrosoftAccount}>
+            <Icon component={Svg} viewBox="0 0 21 21">
+              <rect x="1" y="1" width="9" height="9" fill="#f25022" />
+              <rect x="1" y="11" width="9" height="9" fill="#00a4ef" />
+              <rect x="11" y="1" width="9" height="9" fill="#7fba00" />
+              <rect x="11" y="11" width="9" height="9" fill="#ffb900" />
+            </Icon>
+            <Label>Sign in with Microsoft</Label>
+          </Button>
```
his button should only be here if you don't alread have microsft account linked

zaverbac committed:
refactor(client): ♻️ Remove Google and Microsoft objects from outside functions

Remove Google and Microsoft Auth objects from outside functions and only create them when appropriate function is called

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -1,31 +1,212 @@
+  let microsoftErr = "";
+  let passErr = "";
 
+  function outputErrMsg() {
```
I would prefer if this function had an argument and a return instead of mutating the state

zaverbac committed:
refactor(client): 🔥 Remove confirm password

Remove all code related to confirm password field and checking if passwords match

zaverbac committed:
refactor(client): 💩 In progress of making microsoft and google button not show up when linked

Microsoft and google button need to not show up when account is linked

zaverbac committed:
refactor(client): ♻️ Move provider buttons out of dialog

Move provider buttons out of dialog and onto main account settings page

zaverbac committed:
feat(client): 🐛 Fix provider buttons disappearing at wrong time

Fix provider buttons disappearing when user hasn't linked that provider and showing up when they have

zaverbac committed:
refactor(client): 🔥 Remove unused returns and console.log

Remove unused return true since unreachable code and console.log as it was soley for debugging purposes

zaverbac committed:
refactor(client): ♻️ Change outputErrMsg to take argument and return

Change outputErrMsg function to take an argument of type string and return a string instead of mutating the state

zaverbac committed:
refactor(client): 🔥 Remove hasMicrosoftAccount variable

Remove hasMicrosoftAccount variable due to it not being used anywhere

njumaah commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +40,85 @@ export async function loginAnonymous() {
+  }
+
+  if (user != null) {
+    if (userHasGoogleProvider(user)) {
```
hasGoogleProvider is a better naming convention

nstringham replied:
fixed

njumaah commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +40,85 @@ export async function loginAnonymous() {
+  }
+
+  if (user != null) {
+    if (userHasMicrosoftProvider(user)) {
```
hasMicrosoftProvider is a better naming convention

nstringham replied:
fixed

njumaah commented on client/src/routes/settings/+page.svelte:
```diff
@@ -1,31 +1,198 @@
+    password = "";
+  }
+
+  const googleSvgPaths: { color: string; path: string }[] = [
```
duplicate code from signinbutton component. make another component and insert them in each

njumaah commented on client/src/routes/settings/+page.svelte:
```diff
@@ -48,8 +215,8 @@
-          >Last sign in too long ago.
-          <br />Please Signin and Try Again</Content
+          >Last sign in too long ago. <br />
+          Please Signin and Try Again</Content
```
make a more consistent user message

nstringham replied:
@njumaah is this resolved now? if not can you give an example of what you think would be a better error message or explain what is inconsistent about the current message.

zaverbac replied:
I believe he was just saying be consistent with the wording of sign in as I was saying SignIn on one part and sign in on the other

njumaah commented:
make to to pull from dev, then let me review again before merging

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
refactor(client): ♻️ Change delete account content

Change delete account button content to be more consistent in wording

zaverbac committed:
refactor(client): ♻️ Change userHasGoogleAccount and userHasMicrosoftAccount function names

Change userHasGoogleAccount and userHasMicrosoftAccount to hasGoogleAccount and hasMicrosoftAccount

zaverbac committed:
refactor(client): 🚚 Rename SignInButton.svelte to Account.Svelte

Rename SignInButton.svelte to Account.svelte because it did not accurately depict what it was

zaverbac committed:
refactor(client): 🩹 Add export to userData in Account.svelte

Add export to userData in Account.Svelte cause I accidentally removed it at some point

zaverbac committed:
refactor(client): ♻️ Change Sign In buttons to component

Change Sign In buttons to component to avoid reusing code

zaverbac committed:
Merge branch 'zaverbac/View-Account' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
refactor(client): ♻️ Move functionality of ProviderButtons.svelte to parent components

Move functionality of ProviderButtons.svelte to parent components Account.svelte and Settings

zaverbac committed:
Merge branch 'zaverbac/View-Account' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
refactor(client): ♻️ Move Sign In buttons to ProviderButtons component

Move Sign In buttons to ProviderButtons component and keep functionality on pages

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

njumaah reviewed (dismissed):
lgtm, wait for nates review before merging

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +40,85 @@ export async function loginAnonymous() {
+
+export function hasGoogleProvider(user: User | null) {
+  if (user) {
+    const signInMethods = user.providerData;
```
this would probobly be better as a guard (aka early return)

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +40,85 @@ export async function loginAnonymous() {
+
+export function hasMicrosoftProvider(user: User | null) {
+  if (user) {
+    const signInMethods = user.providerData;
```
this would probobly be better as a guard (aka early return)

zaverbac committed:
refactor(client): ♻️ Add guard clause to hasProvider functions

Add guard clause to hasProvider functions to improve readability

zaverbac committed:
refactor(client): 🔥 Remove if(user) check

Remove if(user) check from hasProvider functions as there is a guard clause checking if user is null

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,78 @@
+      case "microsoft-account-already-linked":
+        return "microsoft account already linked";
+      default:
+        return "An unexpected error has occured";
```
I would rather we just return the message in the else case

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,78 @@
+  let microsoftErr = "";
+  let passErr = "";
+  let deleteErr = "";
+  function outputErrMsg(error: string) {
```
let's give this a more descriptive name

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,78 @@
+      case "google-account-already-linked":
+        return "google account already linked";
+      case "microsoft-account-already-linked":
+        return "microsoft account already linked";
```
these could just be returned from the function instead of being transformed

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,78 @@
       errPrompt = true;
-      errorMsg = err instanceof Error ? err.message : String(err);
+      deleteErr = err instanceof Error ? err.message : String(err);
+      outputErrMsg(deleteErr);
```
you are not doing anything with the error message

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,78 @@
+      window.location.reload();
+    } catch (err) {
+      googleErr = err instanceof Error ? err.message : String(err);
+      outputErrMsg(googleErr);
```
you are not doing anything with the error message

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,78 @@
+      window.location.reload();
+    } catch (err) {
+      microsoftErr = err instanceof Error ? err.message : String(err);
+      outputErrMsg(microsoftErr);
```
you are not doing anything with the error message

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -9,7 +9,7 @@
   import { authStore } from "$stores/auth";
 
-  let userData: UserData | undefined;
+  export let userData: UserData | undefined;
```
I don't think you want this

nstringham commented on client/src/components/ProviderButtons.svelte:
```diff
@@ -0,0 +1,83 @@
+{/if}
+
+<style>
+  :global(#sign-in-with-google),
+  :global(#sign-in-with-microsoft),
+  :global(#sign-in-with-email),
+  :global(#sign-up-with-email) {
+    --mdc-typography-button-text-transform: none;
+    justify-content: start;
+    gap: 4px;
+  }
+
+  :global(#sign-in-with-google),
+  :global(#sign-in-with-microsoft) {
+    --mdc-theme-primary: #ffffff;
+    --mdc-theme-on-primary: #3c4043;
+  }
+
+  @media (prefers-color-scheme: dark) {
+    :global(#sign-in-with-microsoft) {
+      --mdc-theme-primary: #2f2f2f;
+      --mdc-theme-on-primary: #ffffff;
+    }
+  }
```
these styles are duplicated in the settings page and in the account component

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -194,9 +257,107 @@
+    <Button
+      on:click={() => {
+        logOut();
+        goto("/");
```
this should do a page reload like it does for the sign out button in the account dropdown

zaverbac committed:
refactor(client): ♻️ Change outputErrorMsg

Change outputErrorMsg to getErrorMsg and how it sets error messages

zaverbac committed:
refactor(client): 🔥 Remove SignInButtons.svelte

Remove SignInButtons.svelte component as it is redundent code and was replaced with AccountButton component

zaverbac committed:
refactor(client): 🚚 Rename Account.svelte

Rename Account.svelte to AccountButton.svelte as it's a better name

zaverbac committed:
refactor(client): ♻️ Change Account.svelte import in splash page

Change Account import to AccountButton on spalsh page to have it be consistent with component name

zaverbac committed:
fix(client): ♻️ Fix AccountButton imports

Fix AccountButton imports being imported using Account.svelte instead of AccountButton.svelte

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +40,84 @@ export async function loginAnonymous() {
+export function hasMicrosoftProvider(user: User | null) {
+  if (user == null) {
+    return false;
+  }
+  const signInMethods = user.providerData;
```
you put a blank line here in `hasGoogleProvider` but not in `hasMicrosoftProvider`

nstringham replied:
fixed in f7acc7b53c359a49a8d55d432d7edb0e5a160dee

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,77 @@
+
+  let password = "";
+  let errorMsg = "";
+  let googleErr = "";
+  let microsoftErr = "";
+  let passErr = "";
+  let deleteErr = "";
```
`googleErr`, `microsoftErr`, and `deleteErr` are never displayed

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,77 @@
       deleteAccount();
       // If no error
-      errorMsg = "";
-
       goto("/");
```
this should do a reload like logout does

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -194,9 +256,107 @@
-  <Button on:click={() => (showDeletionPrompt = true)}>
-    <Label>Delete Account</Label>
-  </Button>
+  <p id="email">Email: {$user?.email == null ? "Anonymous User" : $user.email}</p>
```
consider using `??`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing

nstringham replied:
resolved in 3ee8bb956cde4dc4ea1245bf9c6cb09f60c1b3b2

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,77 @@
       errPrompt = true;
-      errorMsg = err instanceof Error ? err.message : String(err);
+      deleteErr = getErrorMsg(err);
+      return;
```
you also don't need this `return`

zaverbac committed:
refactor(client): 🔥 Remove returns from functions

Remove return from functions as they were redundant

zaverbac committed:
refactor(client): 🔥 Remove userData export

Remove userData as an export as it wasn't originally an export nor does it need to be one

zaverbac committed:
refactor(client): 🔥 Remove duplicate css

Remove duplicate css in AccountButton component as it was already present in settings page

zaverbac committed:
refactor(client): ♻️ Add page reload

Add page reload

Add page reload to sign out button in account settings page

zaverbac committed:
refactor(client): 🔥 Remove goto

Remove goto("/") as function is already done when logout is called in settings page

zaverbac committed:
refactor(client): ♻️ Add blank line in hasMicrosoftProvider

Add blank line in hasMicrosoftProvider before const signInMethods to make it consistent with hasGoogleProvider function

zaverbac committed:
refactor(client): ♻️ Add page reload

Add page reload in verifyDelete function

zaverbac committed:
refactor(client): ♻️ Add redirect to verifyDelete function

Add window.location.href to redirect user to splash page when account is delete so it also reloads the page

zaverbac committed:
refactor(client): ♻️ Change check if user is anonymous

Change check if user is anonymous and display Anonymous user instead if true by using nullish operator

zaverbac committed:
refactor(client): 🔥 Remove goto

Remove unused goto as user is already being redirected in the logout function

zaverbac committed:
refactor(client): ♻️ Add usage to error variables

Add usage to googleErr, microsoftErr, and deleteErr and displays appropriate error message

zaverbac committed:
Merge branch 'zaverbac/View-Account' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

nstringham commented on client/src/components/AccountButton.svelte:
```diff
@@ -174,28 +174,6 @@
     gap: 12px;
   }
 
-  :global(#sign-in-with-google),
-  :global(#sign-in-with-microsoft),
-  :global(#sign-in-with-email),
-  :global(#sign-up-with-email) {
-    --mdc-typography-button-text-transform: none;
-    justify-content: start;
-    gap: 4px;
-  }
```
we still need the email button code

nstringham commented on client/src/components/ProviderButtons.svelte:
```diff
@@ -0,0 +1,87 @@
+<style>
+  :global(#sign-in-with-google),
+  :global(#sign-in-with-microsoft),
+  :global(#sign-in-with-email),
+  :global(#sign-up-with-email) {
```
there are no email stuff in this component

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -157,6 +221,104 @@
+  <div>
+    <Button
+      on:click={() => {
+        window.location.reload();
```
you don't need to relaod here because logout will already reload

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -157,6 +221,104 @@
+      <div>
+        <Button on:click={linkPassword}>
+          <Label>Set Password</Label>
+        </Button>
```
this button should be an action not in content

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -7,15 +7,25 @@
+  import ProviderButtons from "$components/ProviderButtons.svelte";
 
-  let userData: UserData | undefined = undefined;
+  export let userData: UserData | undefined = undefined;
```
I don't think you want to export this

zaverbac replied:
removed export

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -37,14 +40,78 @@ export async function loginAnonymous() {
     throw new Error("User is not defined.");
   }
 
   deleteUser(user);
```
this needs to be awaited or returned

zaverbac replied:
awaited function

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -83,25 +93,79 @@
   }
 
   let showDeletionPrompt = false;
-  let errorMsg = "";
+  let showOptions = false;
+  let showPassword = false;
   let errPrompt = false;
+  let linkPass = false;
+
+  let password = "";
+  let errorMsg = "";
+  let googleErr = "";
+  let microsoftErr = "";
+  let passErr = "";
+  let deleteErr = "";
 
+  function getErrorMsg(error: unknown): string {
+    let errorMsg = error instanceof Error ? error.message : String(error);
+    switch (errorMsg) {
+      case "Firebase: Password should be at least 6 characters (auth/weak-password).":
+        return "Password should be at least 6 characters";
+      case "Not Signed in":
+        return errorMsg;
+      case "Firebase: Error (auth/popup-closed-by-user).":
+        return "canceled by user";
+      case "Firebase: Error (auth/email-already-in-use).":
+        return "Email already linked to an account";
+      default:
+        return errorMsg;
+    }
+  }
   function verifyDelete() {
     try {
       deleteAccount();
       // If no error
-      errorMsg = "";
-
-      goto("/");
-      return;
+      window.location.href = "/";
```
this doesn't work because you need to await deleteAccount

zaverbac replied:
Added await to both functions

zaverbac committed:
refactor(client): ♻️ Move email css

Move email button css to AccountButton Component where the email butotns reside

zaverbac committed:
refactor(client): 🔥 Remove reload

Remove reload at logout already does the function

zaverbac committed:
refactor(client): ♻️ Move Set Password button

Move Set Password button to Action instead of Content and error displayed on page instead of dialog

zaverbac committed:
feat(client): 💄 Change set password success message color

Add set password success color to green

zaverbac committed:
refactor(client): 🔥 Remove export

Remove export from userData as it shouldnt be exported

zaverbac committed:
refactor(client): ♻️ Add await to deleteUser and verifyDelete

Add await to deleteUser and verifyDelete as redirect doesn't work correctly without doing so

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -35,14 +38,78 @@ export async function loginAnonymous() {
+    throw new Error("Not Signed In");
+  }
+
+  if (user != null) {
+    return updatePassword(user, password);
+  } else {
+    return user;
+  }
```
if you are checking this in the if that throws an error you don't need to check it here

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -35,14 +38,78 @@ export async function loginAnonymous() {
+    throw new Error("Not signed in");
+  }
+
+  if (user != null) {
+    const microsoft = new OAuthProvider("microsoft.com");
+    return linkWithPopup(user, microsoft);
+  }
```
if you are checking this in the if that throws an error you don't need to check it here

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -35,14 +38,78 @@ export async function loginAnonymous() {
+    throw new Error("Not signed in");
+  }
+
+  if (user != null) {
+    const google = new GoogleAuthProvider();
+    return linkWithPopup(user, google);
+  }
```
if you are checking this in the if that throws an error you don't need to check it here

nstringham commented on client/src/routes/settings/+page.svelte:
```diff
@@ -157,6 +221,108 @@
+    </Actions>
+  </Dialog>
+  <!--If user signed in too long ago, redirect them to sign in-->
+  {#if deleteErr !== ""}
```
what does this if statment do?

zaverbac replied:
it opens a dialog that tells the user that they need to sign in again and they click ok and it redirects them back to the sign in  page and logs them out

zaverbac committed:
refactor(client): 🔥 Remove check from functions

Remove check from functions to see if user == null

zaverbac committed:
Merge branch 'zaverbac/View-Account' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -48,38 +48,30 @@ export async function deleteAccount() {
 
 export function linkWithGoogle(user: User | null) {
-  if (user === undefined) {
+  if (user === undefined || user == null) {
```
you can remove `user === undefined ||` because it is not doing anything

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -48,38 +48,30 @@ export async function deleteAccount() {
 
 export function linkWithMicrosoft(user: User | null) {
-  if (user === undefined) {
+  if (user === undefined || user == null) {
```
you can remove `user === undefined ||` because it is not doing anything

nstringham commented on client/src/lib/firebase/auth.ts:
```diff
@@ -48,38 +48,30 @@ export async function deleteAccount() {
 export function linkWithPassword(password: string) {
   const user = auth.currentUser;
-  if (user === undefined) {
+  if (user === undefined || user == null) {
```
you can remove `user === undefined ||` because it is not doing anything

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -80,35 +80,38 @@
       </div>
-      <div class="buttons">
-        <Button on:click={createLobbyHandler} disabled={!nameValidation.valid || waiting} variant="raised">
-          <Label>Create Lobby</Label>
-        </Button>
-        <Button on:click={joinLobbyHandler} disabled={!nameValidation.valid || waiting} variant="raised">
-          <Label>Join Lobby</Label>
-        </Button>
+      <div class="cat-main-buttons">
+        {#if errorMessage !== ""}
+          <p class="error">{errorMessage}</p>
+        {/if}
+        <div class="textbox">
+          <Textfield
+            type="text"
+            label="Display name"
+            bind:value={name}
+            bind:dirty={nameDirty}
+            invalid={nameDirty && !nameValidation.valid}
+            required
+          >
+            <HelperText validationMsg slot="helper">{nameValidation.valid ? "" : nameValidation.reason}</HelperText>
+          </Textfield>
+        </div>
+        <div class="buttons">
+          <Button on:click={createLobbyHandler} disabled={!nameValidation.valid || waiting} variant="raised">
+            <Label>Create Lobby</Label>
+          </Button>
+          <Button on:click={joinLobbyHandler} disabled={!nameValidation.valid || waiting} variant="raised">
+            <Label>Join Lobby</Label>
+          </Button>
+        </div>
```
this seams like a merge conflict thing I don't think you are meaning to do this

zaverbac committed:
refactor(client): 🔥 Remove duplicate dialog

Remove duplicate delete account dialog from settings page

zaverbac committed:
refactor(client): ♻️ Add deleteErr to dialog msg

Add deleteErr to dialog message when user has signed out too long ago

zaverbac committed:
Merge branch 'zaverbac/View-Account' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into zaverbac/View-Account

zaverbac committed:
refactor(client): 🔥 Remove undefine check

Remove undefine check in functions as it is already being checked by user == null

zaverbac committed:
refactor(client): 🔥 Remove unused import

Removed goto import which was being used by deleted deleteAccount dialog

zaverbac committed:
fix(client): 🔥 Manual merge of splash page

Made a manual merge of the splash page since no changes were being suggested

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -8,21 +8,16 @@
   import { displayNameValidator, type UserData } from "$lib/firebase/firestore-types/users";
   import { goto } from "$app/navigation";
   import { authStore } from "$stores/auth";
-
```
Zac, why does this remove every single empty line in this file?

zaverbac replied:
I'm not sure. I didn't remove anything. I know it wasn't showing the splash page on the merge page so I merged it manually by grabbing this from dev and replacing the one I had and changing the SignIn component to AccountButton

zaverbac replied:
Fixed

zaverbac committed:
refactor(client): ♻️ Add lines to splash page
Add lines to splash page after them getting removed somehow

njumaah reviewed (approved):
All acceptance criteria has been taken care of.
